### PR TITLE
Fix dirichlet shape

### LINKED ISSUE #1079: Inconsistency between Dirichlet samples and Categorical argument
There seems to be an inconsistency in the way samples are generated from a Dirichlet, and the way a Categorical interprets its `p` argument. When drawing multiple draws from a k-dimensional Dirichlet, the elements of one draw are along `axis=1` (elements sum to 1 along `axis=1`), while Categorical seems to expect them to be along `axis=0`. I need to take the transpose of the Dirichlet stochastic to get the Categorical to work, without which it throws an error. I am guessing that the same would be true for Multinomial as well, although I haven't checked. This behavior is non-intuitive, especially considering how often Dirichlet is used as a prior for Multinomial/Categorical.

Here is a basic model to reproduce the error (uncomment/comment the appropriate Categorical line)

``` .python
# artificial data
k = 2
ndata = 10
data = np.array([1, 2.2, 2.5, 2.8, 4, 6, 7.2, 7.5, 7.8, 9])

# model
alpha = 0.1 * np.ones((k, ndata))
with pm.Model() as model:
    p = pm.Dirichlet('p', alpha, shape=(k, ndata))
    mu = pm.Normal('mu', mu=5, sd=3, shape=k)
    sd = pm.Uniform('sd', lower=0.1, upper=0.5, shape=k)
    categ = pm.Categorical('categ', p=p.T, shape=(1, ndata))   # works
    # categ = pm.Categorical('categ', p=p, shape=(1, ndata))   # throws an error
    obs = pm.Normal('obs',
                    mu=mu[categ],
                    sd=sd[categ],
                    observed=data)

    step1 = pm.Metropolis(vars=[p, sigma, mu, obs])
    step2 = pm.ElemwiseCategoricalStep(vars=[categ])

    tr = pm.sample(100, step=[step1, step2])
    print tr['p'][0, :]
```

For reference, the trace for `p` looks like this:
`
[[ 0.75233998  0.42171357  0.7272167   0.6370857   0.62833266  0.51940564
   0.74280775  0.7106556   0.58151824  0.30892646]
 [ 0.24766002  0.57828643  0.2727833   0.3629143   0.37166734  0.48059436
   0.25719225  0.2893444   0.41848176  0.69107354]]
`

### PR DESCRIPTION
**Pull request by twiecki**
This is an attempt to fix #1079 using the logic from https://github.com/pymc-devs/pymc3/pull/1236. Unfortunately, while everything else still works, the provided example with the inconsistency to `Categorical` is not fixed and I'm not quite sure why. @taku-y do you have an idea?

Here's the code:

``` python
import numpy as np
import pymc3 as pm

# artificial data
k = 2
ndata = 10
data = np.array([1, 2.2, 2.5, 2.8, 4, 6, 7.2, 7.5, 7.8, 9])

# model
alpha = 0.1 * np.ones((k, ndata))
with pm.Model() as model:
    p = pm.Dirichlet('p', alpha, shape=(k, ndata))
    mu = pm.Normal('mu', mu=5, sd=3, shape=k)
    sd = pm.Uniform('sd', lower=0.1, upper=0.5, shape=k)
    #categ = pm.Categorical('categ', p=p.T, shape=(1, ndata))   # works
    categ = pm.Categorical('categ', p=p, shape=(1, ndata))   # throws an error
    obs = pm.Normal('obs',
                    mu=mu[categ],
                    sd=sd[categ],
                    observed=data)

    step1 = pm.Metropolis(vars=[p, sd, mu, obs])
    step2 = pm.ElemwiseCategorical(vars=[categ])

    tr = pm.sample(100, step=[step1, step2])
    print(tr['p'][0, :])
```

### CODE REVIEW

**Comment by taku-y** [comment]
Thanks for making PR. I will test your code with `pm.Categorical` tomorrow.

File: pymc3/distributions/multivariate.py [hunk 1]
```diff
@@ -103,39 +103,73 @@ def __init__(self, nu, Sigma, mu=None, *args, **kwargs):
...
         log_pdf = gammaln((nu + d)/2.) - 0.5 * (d*tt.log(np.pi*nu) + log_det) - gammaln(nu/2.)
         log_pdf -= 0.5*(nu + d)*tt.log(1 + Q/nu)
-    
+
         return log_pdf
-    
-        
 
 
+class MyDirichlet(Continuous):
```
Comment (taku-y) [inline pymc3/distributions/multivariate.py]: For confirmation, Will `MyDirichlet` be merged to `Dirichlet` in future release?

Comment (twiecki) [inline pymc3/distributions/multivariate.py]: My bad.

**Comment by twiecki** [comment]
The error the above model get's is:

```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-1-2f26faa476d5> in <module>()
     14     sd = pm.Uniform('sd', lower=0.1, upper=0.5, shape=k)
     15     #categ = pm.Categorical('categ', p=p.T, shape=(1, ndata))   # works
---> 16     categ = pm.Categorical('categ', p=p, shape=(1, ndata))   # throws an error
     17     obs = pm.Normal('obs',
     18                     mu=mu[categ],

/home/wiecki/working/projects/pymc/pymc3/distributions/distribution.py in __new__(cls, name, *args, **kwargs)
     23             data = kwargs.pop('observed', None)
     24             dist = cls.dist(*args, **kwargs)
---> 25             return model.Var(name, dist, data)
     26         elif name is None:
     27             return object.__new__(cls)  # for pickle

/home/wiecki/working/projects/pymc/pymc3/model.py in Var(self, name, dist, data)
    281         if data is None:
    282             if getattr(dist, "transform", None) is None:
--> 283                 var = FreeRV(name=name, distribution=dist, model=self)
    284                 self.free_RVs.append(var)
    285             else:

/home/wiecki/working/projects/pymc/pymc3/model.py in __init__(self, type, owner, index, name, distribution, model)
    504             self.tag.test_value = np.ones(
    505                 distribution.shape, distribution.dtype) * distribution.default()
--> 506             self.logp_elemwiset = distribution.logp(self)
    507             self.model = model
    508 

/home/wiecki/working/projects/pymc/pymc3/distributions/discrete.py in logp(self, value)
    396         sumto1 = theano.gradient.zero_grad(tt.le(abs(tt.sum(p, axis=-1) - 1), 1e-5))
    397         if p.ndim > 1:
--> 398             a = tt.log(p[tt.arange(p.shape[0]), value])
    399         else:
    400             a = tt.log(p[value])

/home/wiecki/miniconda3/lib/python3.5/site-packages/theano/tensor/var.py in __getitem__(self, args)
    504                 return self.take(args[axis], axis)
    505             else:
--> 506                 return theano.tensor.subtensor.advanced_subtensor(self, *args)
    507         else:
    508             if numpy.newaxis in args:

/home/wiecki/miniconda3/lib/python3.5/site-packages/theano/gof/op.py in __call__(self, *inputs, **kwargs)
    659                 thunk.outputs = [storage_map[v] for v in node.outputs]
    660 
--> 661                 required = thunk()
    662                 assert not required  # We provided all inputs
    663 

/home/wiecki/miniconda3/lib/python3.5/site-packages/theano/gof/op.py in rval(p, i, o, n)
    903             # default arguments are stored in the closure of `rval`
    904             def rval(p=p, i=node_input_storage, o=node_output_storage, n=node):
--> 905                 r = p(n, [x[0] for x in i], o)
    906                 for o in node.outputs:
    907                     compute_map[o][0] = True

/home/wiecki/miniconda3/lib/python3.5/site-packages/theano/tensor/subtensor.py in perform(self, node, inputs, out_)
   2165         # TODO: in general, we need to re-pack the inputs into a valid
   2166         # index, just like subtensor
-> 2167         out[0] = inputs[0].__getitem__(inputs[1:])
   2168         if (numpy.__version__ <= '1.6.1' and
   2169                 out[0].size != numpy.uint32(out[0].size)):

IndexError: shape mismatch: indexing arrays could not be broadcast together with shapes (2,) (1,10) 
```

**Comment by taku-y** [comment]
Although the problem has not been solved, I report what I noticed. 

I thought of the following syntax.

```
with pm.Model() as model:
    p = pm.Dirichlet('p', alpha, shape=(ndata, k)) # 1st dim samples, 2nd dim elements
```

For test, I set only `p` and `sd` in the model. 

```
alpha = 0.1 * np.ones((ndata, k))
with pm.Model() as model:
    p = pm.Dirichlet('p', alpha, shape=(ndata, k))
    sd = pm.Uniform('sd', lower=0.1, upper=0.5, shape=k)
    step1 = pm.Metropolis(vars=[p, sd])
    tr = pm.sample(100, step=step1)
```

This code works in `master` branch, but failed in `fix_dirichlet_shape` branch. The problem may come from broadcasting. I need to consider how to solve this problem.

**Comment by twiecki** [comment]
Perhaps the problem is actually in `Categorical` which expects the wrong dimension.

**Comment by twiecki** [comment]
@taku-y Seems like I got confused. The first dimension needs to be the size (`ndata`), as you correctly point out. With that, the example indeed works on master.

**Comment by twiecki** [comment]
All things considered, it now seems to me that the shape handling of Dirichlet and Categorical is indeed correct, but I'm not sure. What are your thoughts?

**Comment by fonnesbeck** [comment]
We should be able to infer k for multivariate nodes from the shape of the arguments, should we not? For example, a Dirichlet always has a vector-valued parameter, so we should be able to say:

```
with Model() as model:
    p = Dirichlet('p',  alpha, shape=ndata)
```

and `k` is assumed to be the `shape[0]` of `alpha`.  So each draw of `p` will have shape `(k,ndata)`.

**Comment by taku-y** [comment]
@twiecki Sorry for the confusion. 

```
alpha = 0.1 * np.ones((ndata, k))
with pm.Model() as model:
    p = pm.Dirichlet('p', alpha, shape=(ndata, k))
    sd = pm.Uniform('sd', lower=0.1, upper=0.5, shape=k)
    step1 = pm.Metropolis(vars=[p, sd])
    tr = pm.sample(100, step=step1)
```

The above code works on `master`, but failed on `fix_dirichlet_shape`. 

In the above code, `p` and `sd` are independent with each other. Nevertheless, I got an error of `The broadcastable pattern of the input is incorrect for this op. Expected (False, True), got (False, False).`  It raised when calling `join_nonshared_inputs([logp], vars, shared)`, where Theano variables are concatenated. This problem is not related to inconsistency of `Dirichlet` and `Categorical`. It may be an issue of usage of Theano.

**Comment by taku-y** [comment]
The stack trace of the error is like this. 

```
TypeErrorTraceback (most recent call last)
<ipython-input-1-3c96996c551e> in <module>()
     15     p = pm.Dirichlet('p', alpha, shape=(ndata, k))
     16     sd = pm.Uniform('sd', lower=0.1, upper=0.5, shape=k)
---> 17     step1 = pm.Metropolis(vars=[p, sd])
     18     tr = pm.sample(100, step=step1)

/home/jovyan/work/git/github/taku-y/pymc3/pymc3/step_methods/arraystep.py in __new__(cls, *args, **kwargs)
     56                 # If we don't return the instance we have to manually
     57                 # call __init__
---> 58                 step.__init__([var], *args, **kwargs)
     59                 # Hack for creating the class correctly when unpickling.
     60                 step.__newargs = ([var], ) + args, kwargs

/home/jovyan/work/git/github/taku-y/pymc3/pymc3/step_methods/metropolis.py in __init__(self, vars, S, proposal_dist, scaling, tune, tune_interval, model, **kwargs)
     99 
    100         shared = make_shared_replacements(vars, model)
--> 101         self.delta_logp = delta_logp(model.logpt, vars, shared)
    102         super(Metropolis, self).__init__(vars, shared)
    103 

/home/jovyan/work/git/github/taku-y/pymc3/pymc3/step_methods/metropolis.py in delta_logp(logp, vars, shared)
    287 
    288 def delta_logp(logp, vars, shared):
--> 289     [logp0], inarray0 = join_nonshared_inputs([logp], vars, shared)
    290 
    291     tensor_type = inarray0.type

/home/jovyan/work/git/github/taku-y/pymc3/pymc3/theanof.py in join_nonshared_inputs(xs, vars, shared, make_shared)
    189     replace.update(shared)
    190 
--> 191     xs_special = [theano.clone(x, replace, strict=False) for x in xs]
    192     return xs_special, inarray
    193 

/home/jovyan/work/git/github/taku-y/pymc3/pymc3/theanof.py in <listcomp>(.0)
    189     replace.update(shared)
    190 
--> 191     xs_special = [theano.clone(x, replace, strict=False) for x in xs]
    192     return xs_special, inarray
    193 

/opt/conda/lib/python3.5/site-packages/theano/scan_module/scan_utils.py in clone(output, replace, strict, share_inputs, copy_inputs)
    252                                         [],
    253                                         strict,
--> 254                                         share_inputs)
    255 
    256     return outs

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in rebuild_collect_shared(outputs, inputs, replace, updates, rebuild_strict, copy_inputs_over, no_default_updates)
    229     else:
    230         if isinstance(outputs, Variable):
--> 231             cloned_v = clone_v_get_shared_updates(outputs, copy_inputs_over)
    232             cloned_outputs = cloned_v
    233             # computed_list.append(cloned_v)

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     90             if owner not in clone_d:
     91                 for i in owner.inputs:
---> 92                     clone_v_get_shared_updates(i, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(

/opt/conda/lib/python3.5/site-packages/theano/compile/pfunc.py in clone_v_get_shared_updates(v, copy_inputs_over)
     93 
     94                 clone_d[owner] = owner.clone_with_new_inputs(
---> 95                     [clone_d[i] for i in owner.inputs], strict=rebuild_strict)
     96                 for old_o, new_o in zip(owner.outputs, clone_d[owner].outputs):
     97                     clone_d.setdefault(old_o, new_o)

/opt/conda/lib/python3.5/site-packages/theano/gof/graph.py in clone_with_new_inputs(self, inputs, strict)
    244                     remake_node = True
    245         if remake_node:
--> 246             new_node = self.op.make_node(*new_inputs)
    247             new_node.tag = copy(self.tag).__update__(new_node.tag)
    248         else:

/opt/conda/lib/python3.5/site-packages/theano/tensor/elemwise.py in make_node(self, _input)
    200                         "The broadcastable pattern of the "
    201                         "input is incorrect for this op. Expected %s, got %s."
--> 202                         % (self.input_broadcastable, ib)))
    203                 # else, expected == b or expected is False and b is True
    204                 # Both case are good.

TypeError: The broadcastable pattern of the input is incorrect for this op. Expected (False, True), got (False, False).
```

**Comment by taku-y** [comment]
@fonnesbeck 

> We should be able to infer k for multivariate nodes from the shape of the arguments, should we not? 

I prefer to infer `k` from `a.shape`; if `a.ndim` is 1, `k` is `a.shape[0]`. Otherwise, `k` is `a.shape[-1]` (https://github.com/pymc-devs/pymc3/pull/1125#issuecomment-222850131).

**Comment by brandonwillard** [comment]
@fonnesbeck, @taku-y: #1125 has some cursory changes that reflect the ideas discussed here, as well as a few possible [implementation details](https://github.com/brandonwillard/pymc3/blob/dist-shape-refactor/pymc3/distributions/multivariate.py#L293).

**Comment by taku-y** [comment]
I fixed an error mentioned above. For the record, I describe the problem and how to fix it. Example notebook is https://gist.github.com/taku-y/0d5965a184066699cfdc22d9e71d032b. I pushed a commit with the fix (https://github.com/taku-y/pymc3/tree/fix_dirichlet_shape_taku-y). 

```
# artificial data
# k = 2 # error
k = 3
ndata = 10
data = np.array([1, 2.2, 2.5, 2.8, 4, 6, 7.2, 7.5, 7.8, 9])

# model
alpha = 0.1 * np.ones((ndata, k))
with pm.Model() as model:
    p = pm.Dirichlet('p', alpha, shape=(ndata, k))
    mu = pm.Normal('mu', mu=5, sd=3, shape=k)
    sd = pm.Uniform('sd', lower=0.1, upper=0.5, shape=k)
    categ = pm.Categorical('categ', p=p, shape=ndata)
    obs = pm.Normal('obs',
                   mu=mu[categ],
                   sd=sd[categ],
                   observed=data)

    step1 = pm.Metropolis(vars=[p, mu, sd, obs])
    step2 = pm.ElemwiseCategorical(vars=[categ])

    tr = pm.sample(100, step=[step1, step2])
```

The above code ran with `k=3`, but threw an error in calling `theano.clone()` with `k=2`. For the later case, the shape of `p_stickbreaking_`, the stick-breaking transform of `p`, was `(10, 1)` (`=(ndata, k-1)`). Since the shape included `1`, PyMC3 inferred the type of `p_stickbreaking_` as `TensorType(float64, col)` by the following code in `distribution.py`. 

```
def TensorType(dtype, shape):
    return tt.TensorType(str(dtype), np.atleast_1d(shape) == 1)
```

However, when making the theano graph with replacing `p_stickbreaking_` for some computation, PyMC3 tried to replace it with a tensor of type `TensorType(float64, matrix)`, and resulted in the error. On the other hand, when `k=3`, the shape of `p_stickbreaking_` is inferred as `TensorType(float64, matrix)`, because the shape of `p_stickbreaking_` was `(10, 2)`.

To solve this problem, I added the following code in the constructor of `TransformedDistribution`. 

```
if transform.name == 'stickbreaking':
    b = np.hstack(((np.atleast_1d(self.shape) == 1)[:-1], False))
    self.type = tt.TensorType(v.dtype, b) # force the last dim not broadcastable
```

If I understand Theano correctly, `col` means the second axis (`axis=1`) is broadcastable. However, under the design of the shape of random variables considered in this issue, the second axis corresponds to the dimension of the random vectors, so it should not be broadcastable. For this reason, I think the above fix can be applied to other transformations (`if transform.name == 'stickbreaking':` can be removed).

**Comment by twiecki** [comment]
@taku-y Glad you figured it out, thanks. Btw you can just commit directly to this branch as you have write access now. Want to add that?

**Comment by taku-y** [comment]
Thanks for telling me that, I did not know. I will add it with other fixes for passing tests.

**Comment by twiecki** [comment]
Thanks @taku-y, looks good to me, what do you think?
@fonnesbeck?

**Comment by fonnesbeck** [comment]
So, instead of:

```
p = pm.Dirichlet('p', a=a, shape=k)
```

in `gaussian_mixture_model`, we should be able to:

```
p = pm.Dirichlet('p', a=a)
```

since shape is being inferred from `a`, however the `shape` of `a` is a Theano tensor so this does not work (since shape is tested as an `int`).

**Comment by twiecki** [comment]
@fonnesbeck This is not about shape inference, but about inconsistent dimension handling in `Dirichlet`.

**Comment by taku-y** [comment]
@twiecki I think it's ready to merge. 

@fonnesbeck I agree with you, but I'd say some base classes like `Continuous` need to be modified for the case of being given the parameter `a` as Theano tensor. It should be discussed in #1125.

**Pull request merged**